Hai, kita tau kalau integral itu sering dibilang sebagai luas di bawah kurva. Dengan rumus kayak

$$\int_{x_i}^{x_f}x^n\,dx = (\frac{1}{n+1}x_f^{n+1} - \frac{1}{n+1}x_i^{n+1}) + c$$

kita bisa menghitung luas di bawah kurva fungsi apapun dengan akurat. Mungkin dari kalian ada yang mikir, "mana unsur luasnya dalam rumusnya?"

Kita bisa bayangin kalau simbol integral itu penjumlahan, semacam simbol $\Sigma$. Kalau misal kita punya grafik kayak gini, kita bisa potong rata sepanjang misalnya $dx$, terus kita tarik garis sepanjang $f(x)$ secara vertikal. Bisa dibilang kalau persegi-persegi ini kita jumlahin kita dapet integral dari fungsinya. Dengan konsep tadi, kalau kita potong sampai kecil banget, hasilnya bakal akurat banget kan? Di sini kita akan coba untuk menyelesaikan integral dengan cara ini.

Buat contoh, kita mau cari integral dari $x^2$ dari 1 sampai 4. Pertama, kita mau tau cara bagi rata angka dari 1 sampai 4.

In [ ]:
a = 1
b = 4
n = 5
dx = (b-a)/n

for i in range(n+1):
    print(a + dx*i)

1.0
1.6
2.2
2.8
3.4
4.0


Nah, kita sudah punya titik-titiknya. Sekarang kita lakuin tarik garis dan hitung luas dari persegi yang terbentuk

In [ ]:
def f(x):
    return x**2

total = 0
for i in range(n):
    total += dx * f(a + dx*i)

print(total)

16.68


Hasilnya 16.68 menurut ini, coba kita cek sendiri.

$$\int_1^4 x^2 \,dx = \frac{1}{3}4^3 - \frac{1}{3}1^3 = 21$$

keliatannya kurang bener, ini karena kekosongan dari pendekatan persegi menyebabkan hasilnya kurang dari yang seharusnya. Kalau `n` kita banyakin, hasilnya harusnya semakin mendekati.

In [ ]:
n = 100000
dx = (b-a)/n

total = 0
for i in range(n):
    total += dx * f(a + dx*i)

print(total)

20.99977500045012


Sudah lumayan mendekati sekarang. Sebenarnya metode ini ada dua versi, satu yang kekurangan yang sudah kita coba dan versi yang kelebihan. Untuk yang versi kelebihan, kita tinggal modif satu hal aja dari kodenya. Sekarang, harusnya kita mendapatkan hasil lebih besar dari yang versi kekurangan.

In [ ]:
n = 100000
dx = (b-a)/n

total = 0
for i in range(n):
    total += dx * f(a + dx*(i+1))

print(total)

21.000225000450122


Bisa gak kita cari metode lain yang lebih akurat? Tentu bisa, mungkin ada yang sudah kepikiran, gimana kalau bentuknya trapesium? Caranya masih sama, cuma beda di rumus mencari luas aja. Rumus luas trapesium itu 

$$\frac{1}{2}(a+b)h$$

di sini, $a$ dan $b$ itu tinggi dari fungsinya dan $h$ itu adalah $dx$ kita


In [ ]:
n = 5
dx = (b-a)/n

total = 0
for i in range(n):
    x1 = a + dx*i
    x2 = a + dx*(i+1)
    total += 1/2 * (f(x1) + f(x2)) * dx

print(total)

21.18


Kita coba dengan `n = 5` dan hasilnya sudah mendekati, di sini kita bisa liat dengan metode yang lebih bagus, hasilnya lebih cepat mendekati yang sebenarnya. Buat yang pengen tau, ini hasilnya kalau 100000 iterasi.

In [ ]:
n = 100000
dx = (b-a)/n

total = 0
for i in range(n):
    x1 = a + dx*i
    x2 = a + dx*(i+1)
    total += 1/2 * (f(x1) + f(x2)) * dx

print(total)

21.000000000449916


Buat nyicip metode apa lagi yang ada, ini aku kasih lihat metode Simpson 1/3. 

$$
\int_a^b f(x) dx \approx \frac{b-a}{3n} \left( f(x_0) + 4 \sum_{i = 1,3,5,...}^{n-1} f(x_i) + 2 \sum_{i = 2,4,6,...}^{n-2} f(x_i) + f(x_{n})\right)
$$
dengan $x_0 = a$ dan $x_i = a + ih$ dan nilai n harus genap

In [ ]:
n = 6
dx = (b-a)/n

total = f(a) + f(a + n*dx)

for i in range(1, n, 2):
  total += 4*f(a + i*dx)
  total += 2*f(a + (i+1)*dx)

total *= dx/3

print(total)

26.333333333333332


Kok pas dicoba `n = 5` hasilnya lebih jelek dari yang trapesium? Masing-masing metode punya kelebihan dan kekurangannya masing-masing. Kita coba pakai metode Simpson untuk menghitung integral dari fungsi $\sin(x)$ dari 0 sampai $\frac{\pi}{2}$ dan kita bandingkan dengan metode trapesium

Hasil yang seharusnya didapatkan adalah
$$\int_0^{\frac{\pi}{2}} \sin x\,dx = -\cos{\frac{\pi}{2}} -(-\cos{0}) = 1$$

In [ ]:
from math import sin, pi, exp
def f(x):
    return sin(x)

a = 0
b = pi
n = 4
dx = (b-a)/n

total = f(a) + f(a + n*dx)

for i in range(1, n, 2):
  total += 4*f(a + i*dx)
  total += 2*f(a + (i+1)*dx)

total *= dx/3

print(total)
print(total-2)

2.0045597549844207
0.0045597549844207386


In [ ]:
dx = (b-a)/n

total = 0
for i in range(n):
    x1 = a + dx*i
    x2 = a + dx*(i+1)
    total += 1/2 * (f(x1) + f(x2)) * dx

print(total)
print(total-2)

1.8961188979370398
-0.1038811020629602


Dengan `n = 4`, metode Simpson 1/3 sudah bisa mendekati hasil integralnya.

Beberapa metode lainnya yang aku tau itu
1. Metoda Simpson 3/8, sama kayak Simpson 1/3 tapi lebih akurat dan
1. Monte carlo, ini unik, kusaranin untuk dipelajari
3. Ada juga pakai SciPy.integrate.quad

Semoga berguna, bhay!